src/Orleans.Transactions/State/TransactionalState.cs

```diff
@@ -0,0 +1,516 @@
+                            break;
+                        }
+
+                        highestKey = key;
+                    }
+
+                    if (this.log.Count != 0)
+                    {
+                        List<KeyValuePair<long, LogRecord<TState>>> records = this.log.TakeWhile(kvp => kvp.Key < highestKey).ToList();
+                        records.ForEach(kvp => this.log.Remove(kvp.Key));
+                    }
+
+                    await Persist(transactionId, this.writeLowerBound);
+                    return true;
+                });
+            }
+        }
+        #endregion ITransactionalResource
+
+
+        /// <summary>
+        /// Find the appropriate version of the state to serve for this transaction.
+        /// We enforce reads in transaction id order, hence we find the version written by the highest 
+        /// transaction less than or equal to this one
+        /// </summary>
+        private bool TryGetVersion(long transactionId, out TState readState, out TransactionalResourceVersion readVersion)
+        {
+            readState = this.value;
+            readVersion = this.version;
+            bool versionAvailable = this.version.TransactionId <= transactionId;
+
+            LogRecord<TState> logRecord = null;
+            foreach (KeyValuePair<long, LogRecord<TState>> kvp in this.log)
+            {
+                if (kvp.Key > transactionId)
+                {
+                    break;
+                }
+                logRecord = kvp.Value;
+            }
+
+            if (logRecord == null) return versionAvailable;
+
+            readState = logRecord.NewVal;
+            readVersion = logRecord.Version;
+
+            return true;
+        }
+
+        /// <summary>
+        /// Transactional Read procedure.
+        /// </summary>
+        private TState GetState()
+        {
+
+            var info = TransactionContext.GetTransactionInfo();
+
+            if (this.logger.IsVerbose2) this.logger.Verbose2("Read {0}", info);
+
+            Restore();
+
+            if (this.transactionCopy.TryGetValue(info.TransactionId, out TState state))
+            {
+                return state;
+            }
+
+            if (!TryGetVersion(info.TransactionId, out TState readState, out TransactionalResourceVersion readVersion))
+            {
+                // This can only happen if old versions are gone due to checkpointing.
+                throw new OrleansTransactionVersionDeletedException(info.TransactionId);
+            }
+
+            if (info.IsReadOnly && readVersion.TransactionId > this.stableVersion)
+            {
+                throw new OrleansTransactionUnstableVersionException(info.TransactionId);
+            }
+
+            info.RecordRead(transactionalResource, readVersion, this.storage.State.StableVersion);
+
+            writeLowerBound = Math.Max(writeLowerBound, info.TransactionId - 1);
+
+            TState copy = this.copier.DeepCopy(readState);
```

Comment:
No change needed here, but if we know that `readState` was created for a read-only tx then we could avoid the deep copy. Just an idea.

Reply:
There is still some wiring here related to readonly, but readonly transactions had issues and is not fully supported yet.  Readonly optimization will be revisited after primary path is baked.  If we can't get it right it will be removed entirely.